ai-catalog/samples/genai-image-description/src/main/java/com/android/ai/samples/genai_image_description/GenAIImageDescriptionViewModel.kt

```diff
@@ -15,89 +15,127 @@
-                    _resultGenerated.value =
-                        context.getString(R.string.genai_image_description_not_available)
-                    return@launch
-                }
-
-                // If feature is downloadable, making an inference call will automatically start
-                // the downloading process.
-                // If feature is downloading, the inference request will automatically execute after
-                // the feature has been downloaded.
-                // Alternatively, you can call imageDescriber.downloadFeature() to monitor the
-                // progress of the download.
-                if (featureStatus == FeatureStatus.DOWNLOADABLE ||
-                    featureStatus == FeatureStatus.DOWNLOADING
-                ) {
-                    _resultGenerated.value =
-                        context.getString(R.string.genai_image_description_downloading)
-                }
-
-                val bitmap = MediaStore.Images.Media.getBitmap(context.contentResolver, imageUri)
-                val request = ImageDescriptionRequest.builder(bitmap).build()
-                imageDescriber.runInference(request) { newText ->
-                    if (_resultGenerated.value ==
-                        context.getString(R.string.genai_image_description_downloading)
-                    ) {
-                        clearGeneratedText()
-                    }
-                    _resultGenerated.value += newText
-                }
+        imageDescriptionJob = viewModelScope.launch {
+            var featureStatus = FeatureStatus.UNAVAILABLE
+
+            try {
+                _uiState.value = GenAIImageDescriptionUiState.CheckingFeatureStatus
+                featureStatus = imageDescriber.checkFeatureStatus().await()
+            } catch (error: Exception) {
+                _uiState.value = GenAIImageDescriptionUiState.Error(R.string.image_desc_feature_check_fail)
+                Log.e("GenAIImageDesc", "Error checking feature status", error)
+            }
+
+            if (featureStatus == FeatureStatus.UNAVAILABLE) {
+                _uiState.value = GenAIImageDescriptionUiState.Error(R.string.genai_image_description_not_available)
                 return@launch
             }
+
+            if (featureStatus == FeatureStatus.DOWNLOADABLE || featureStatus == FeatureStatus.DOWNLOADING) {
+                imageDescriber.downloadFeature(
+                    object : DownloadCallback {
+                        override fun onDownloadStarted(bytesToDownload: Long) {
+                            _uiState.value = GenAIImageDescriptionUiState.DownloadingFeature(bytesToDownload, 0)
+                        }
+
+                        override fun onDownloadProgress(bytesDownloaded: Long) {
+                            _uiState.update {
+                                (it as? GenAIImageDescriptionUiState.DownloadingFeature)?.copy(bytesDownloaded = bytesDownloaded) ?: it
```

Comment:
I wish I could test this somehow, but I don't think the download actually gets cancelled by the coroutine. That means this might still fire and mess with your UI state after you have set it back to initial from dismissing the bottom sheet.

The easiest way to fix is just to not support cancellation and remove the bottom sheet

Reply:
okay yeah i tested it, it gets really wonky if i hide the bottom sheet while it's still downloading.

removed the bottom sheet implementation